### PR TITLE
[WB-7468] [WB-7692] Settings refactor

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-7468
Fixes WB-7692
...and a ton of other stuff

I apologize for the size of this PR and the length of the description below :)

Description
-----------
Context: currently, settings are handled in a very complex manner, making it somewhat challenging to add new features and to maintain the existing ones. 

The main purpose of this PR (which is likely only the first one in a series of PRs) is to refactor the `wandb.sdk.wandb_settings.Settings` class to provide:
- Consistency, tight control, and encapsulation: implement general interfaces for settings.
  - Enable preprocessing, validation, and runtime hooks (e.g. a setting can depend on another setting) for individual settings.
  - Implement and enforce source priority logic in a consistent matter. Brake settings into policies and non-policy ones.
- Compactness

Having addressed the above, I focused on making sure that nothing is broken across the client. 

It's been a lot of fun working on this, and I expect even more fun from the review process :) 

Implementation details
----------------------
This PR adds the `Property` class that is utilized by the `Settings`.

`Property`: a class to represent attributes (individual settings) of the Settings object.
- Encapsulates the logic of how to preprocess and validate values of settings throughout the lifetime of a class instance.
- Allows for runtime modification of settings with hooks, e.g. in the case when a setting depends on another setting.
- The `update()` method is used to update the value of a setting. Enforces source priority logic when updating values.
- The `is_policy` attribute determines the source priority when updating the property value. E.g. if `is_policy` is `True`, the smallest `Source` value takes precedence.
- Ability to freeze/unfreeze.

Here's a basic example (for more examples, see `tests/wandb_settings_test.py`)

```python
from wandb.sdk.wandb_settings import Property, Source


def uses_https(x):
    if not x.startswith("https"):
        raise ValueError("Must use https")
    return True

base_url = Property(
    name="base_url",
    value="https://wandb.com/",
    preprocessor=lambda x: x.rstrip("/"),
    validator=[lambda x: isinstance(x, str), uses_https],
    source=Source.BASE,
)

endpoint = Property(
    name="endpoint",
    value="site",
    validator=lambda x: isinstance(x, str),
    hook=lambda x: "/".join([base_url.value, x]),
    source=Source.BASE,
)
```

```python
>>> print(base_url)  # note the stripped "/"
'https://wandb.com'
>>> print(endpoint)  # note the runtime hook
'https://wandb.com/site'
>>> print(endpoint._value)  # raw value
'site'
>>> base_url.update(value="https://wandb.ai/", source=Source.INIT)
>>> print(endpoint)  # valid update with a higher priority source
'https://wandb.ai/site'
>>> base_url.update(value="http://wandb.ai/")  # invalid value - second validator will raise exception
ValueError: Must use https
>>> base_url.update(value="https://wandb.dev", source=Source.USER)
>>> print(endpoint)  # valid value from a lower priority source has no effect
'https://wandb.ai/site'
```

`Settings`: a class to that uses `Property` objects to represent configurable settings.

- The code is supposed to be self-documented -- see `wandb/sdk/wandb_settings.py` :)
- Clearly and compactly defines all individual settings, their default values, preprocessors, validators, and runtime hooks as well as whether or not they are treated as policies.
- Provides a mechanism to update settings specifying source (which abides the corresponding Property source logic) via `Settings.update()`. Direct attribute assignment is not allowed.
- Careful Settings object copying.
- Mapping interface.
- Exposes `attribute.value` if attribute is a `Property`.
- Ability to freeze/unfreeze.
- `Settings.make_static()` method that we can use to replace `StaticSettings`.
- Adapted/reworked convenience methods to apply settings originating from different source.

Other stuff in this PR
-------

- Fix WB-7468 (code saving and wandb.login()).
- Fix WB-7692 (drop some of the deps back-ported to py2).
- A note on using a debugger in `CONTRIBUTING.md`.
- Configurable `'{"code_saving_enabled": true}'` in `mock_server` ctx.
- Improved test suite robustness (see below).
- Expose run settings via a public attribute `Run.settings`. We should also start saving them by default with each run as it would provide a lot of useful information that is not otherwise captured.

Testing
-------
Numerous (>100) tests, both settings-specific and not, were added or fixed.
- No more sloppiness in the way the settings are treated in the tests :)
- Improved isolation for numerous tests that were leaking state through filesystem and env variables and potentially affecting other tests (in a flaky way) executed by the same `pytest` worker.

Some notes
-------
When designing this, I considered external libraries `attrs` and `pydantic` as potentially useful. Neither seemed like a good option for us:
- `pydantic` generates scary code behind the scenes.
- `attrs` only ensures conversion and validation at init.
- Neither have runtime hooks/converters/validation.
- No easy way to implement tracking and updating values based on source priorities.

Status as of 1/11/2022
------
- A single functional test is failing: disabled code saving. This happens bc the mock server sets `{“code_saving_enabled”: True}` in the user settings by default, which, since now code saving setting is a policy, has higher priority than the ENV variable that the test sets. In unit tests, I made that configurable and so I do `live_mock_server.set_ctx({"code_saving_enabled": None})` before running the corresponding tests — now I am working on making this possible with `yea`. Shouldn't take too long, but I think the review can proceed in the meantime.
- Several tests on Windows are failing -- need to fix them, but I also think it's not a road block for the review.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -173,9 +171,13 @@ def test_dir(test_name):
...
 @pytest.fixture
 def disable_git_save():
+    env_var_is_set = "WANDB_DISABLE_GIT" in os.environ
     os.environ["WANDB_DISABLE_GIT"] = "true"
     yield
-    os.environ["WANDB_DISABLE_GIT"] = "false"
+    if env_var_is_set:
+        os.environ["WANDB_DISABLE_GIT"] = "false"
+    else:
+        del os.environ["WANDB_DISABLE_GIT"]
```
Comment (kptkin) [inline tests/conftest.py:180]: should we use pytest mock instead?

File: tests/conftest.py [hunk 2]
```diff
@@ -232,21 +234,23 @@ def test_settings(test_dir, mocker, live_mock_server):
...
-        _start_datetime=datetime.datetime.now(),
+        save_code=False,
     )
     yield settings
-    # Just incase someone forgets to join in tests
+    # Just in case someone forgets to join in tests. ...well, don't!
     if wandb.run is not None:
         wandb.run.finish()
+    # if wandb.wandb_sdk.wandb_setup._WandbSetup.instance is not None:
+    #     wandb.wandb_sdk.wandb_setup._WandbSetup.instance = None
```
Comment (kptkin) [inline tests/conftest.py:253]: commented out code

File: tests/conftest.py [hunk 3]
```diff
@@ -255,6 +259,8 @@ def mocked_run(runner, test_settings):
     run = wandb.wandb_sdk.wandb_run.Run(settings=test_settings)
     run._set_backend(MagicMock())
     yield run
+    # if wandb.run is not None:
+    #     wandb.run.finish()
```
Comment (kptkin) [inline tests/conftest.py:263]: commented out code

File: tests/conftest.py [hunk 4]
```diff
@@ -471,11 +477,11 @@ def wandb_init_run(request, runner, mocker, mock_server):
         wandb.wandb_sdk.wandb_setup._WandbSetup._instance = None
         mocker.patch("wandb.wandb_sdk.wandb_init.Backend", utils.BackendMock)
         run = wandb.init(
-            settings=wandb.Settings(console="off", mode="offline", _except_exit=False),
+            settings=dict(console="off", mode="offline", _except_exit=False),
             **args["wandb_init"],
         )
         yield run
-        wandb.join()
+        wandb.finish()
```
Comment (kptkin) [inline tests/conftest.py:484]: nice

File: tests/integrations/test_keras.py [hunk 5]
```diff
@@ -89,6 +86,7 @@ def test_basic_keras(dummy_model, dummy_data, wandb_init_run):
     # NOTE: backend mock doesnt copy history into summary (happens in internal process)
     # assert wandb.run._backend.summary["loss"] > 0
     assert len(graph_json(wandb.run)["nodes"]) == 3
+    wandb.finish()
```
Comment (kptkin) [inline tests/integrations/test_keras.py:89]: why do you need to add finish here?

File: tests/integrations/test_keras.py [hunk 6]
```diff
@@ -901,16 +919,19 @@ def test_data_logger_pred_inferred_proc_no_classes(test_settings, live_mock_serv
...
+    dummy_model.fit(
+        *dummy_data,
+        epochs=2,
+        batch_size=36,
+        validation_data=dummy_data,
+        callbacks=[WandbCallback(log_evaluation=True)]
+    )
+    assert wandb.run.summary["validation_predictions"] is not None
+    assert wandb.run.summary["validation_predictions"]["artifact_path"] is not None
+    assert wandb.run.summary["validation_predictions"]["_type"] == "table-file"
```
Comment (kptkin) [inline tests/integrations/test_keras.py:936]: Nice!

File: tests/test_redir.py [hunk 7]
```diff
@@ -235,18 +231,23 @@ def test_run_with_console_redirect(console_settings, capfd):
 
         print("\n" * 1000)
         print("---------------")
+        time.sleep(1)
```
Comment (kptkin) [inline tests/test_redir.py:234]: why do you need to add this?

File: wandb/sdk/wandb_setup.py [hunk 8]
```diff
@@ -161,32 +174,31 @@ def _get_logger(self):
...
 
-    def _load_user_settings(self, settings=None):
+    def _load_user_settings(self, settings=None) -> Optional[Dict[str, Any]]:
         if self._server is None:
-            self._load_viewer()
+            self._load_viewer(settings=settings)
+
+        # offline?
+        if self._server is None:
+            return None
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:198]: Can be under the same condition. Why do you need to change this? How did it use to work?

File: wandb/sdk/wandb_run.py [hunk 9]
```diff
@@ -1,8 +1,6 @@
 #
 # -*- coding: utf-8 -*-
```
Comment (kptkin) [inline wandb/sdk/wandb_run.py:2]: probably can remove this one as well... not needed in python 3 since the default encoding is already utf-8

File: wandb/sdk/wandb_init.py [hunk 10]
```diff
@@ -171,47 +178,55 @@ def setup(self, kwargs) -> None:
...
 
         if not settings._noop:
             self._log_setup(settings)
 
             if settings._jupyter:
                 self._jupyter_setup(settings)
 
-        self.settings = settings.freeze()
+        self.settings = settings
+        # self.settings.freeze()
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:229]: commented out code

File: wandb/sdk/wandb_init.py [hunk 11]
```diff
@@ -171,47 +178,55 @@ def setup(self, kwargs) -> None:
...
 
         # handle login related parameters as these are applied to global state
-        anonymous = kwargs.pop("anonymous", None)
-        force = kwargs.pop("force", None)
-
-        # TODO: move above parameters into apply_init_login
-        settings._apply_init_login(kwargs)
+        init_settings = dict()
+        for key in ["anonymous", "force", "mode", "resume"]:
+            if key in kwargs and kwargs[key] is not None:
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:183]:
```suggestion
            if kwargs.get(key, None) is not None:
```

File: wandb/sdk/wandb_init.py [hunk 12]
```diff
@@ -107,27 +108,33 @@ def setup(self, kwargs) -> None:
...
-
-        self._reporter = reporting.setup_reporter(
-            settings=settings.duplicate().freeze()
-        )
-
-        sm_config: Dict = (
-            {} if settings.sagemaker_disable else sagemaker.parse_sm_config()
+        if settings_param is not None:
+            if isinstance(settings_param, Settings):
+                # todo: check the logic here. this _only_ comes up in tests?
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:115]: did you resolve this TODO?

File: wandb/sdk/lib/reporting.py [hunk 13]
```diff
@@ -90,8 +90,9 @@ def __getattr__(self, name):
 
 
 def setup_reporter(settings):
-    if not settings.is_frozen():
-        logging.error("internal issue: settings not frozen")
+    # fixme: why?
+    # if not settings.is_frozen():
+    #     logging.error("internal issue: settings not frozen")
```
Comment (kptkin) [inline wandb/sdk/lib/reporting.py:95]: did you resolve this one?

File: tests/wandb_integration_test.py [hunk 14]
```diff
@@ -63,134 +48,141 @@ def test_resume_allow_success(live_mock_server, test_settings):
...
+        run.finish()
 
-    with pytest.raises(wandb.errors.UsageError):
         run = wandb.init(
             reinit=True,
             resume=True,
             settings=test_settings,
             config=config,
-            config_exclude_keys=("bar",),
             config_include_keys=("bar",),
```
Comment (kptkin) [inline tests/wandb_integration_test.py:164]: haha

File: tests/integrations/test_keras.py (hunk 5 above)
Comment (dmitryduev) [reply tests/integrations/test_keras.py:89]: I think we should explicitly finish things up and not be relying on the `test_settings` fixture that will do it at teardown.

File: tests/integrations/test_keras.py (hunk 6 above)
Comment (dmitryduev) [reply tests/integrations/test_keras.py:936]: Actually, I am not sure what happened here. It used to be `artifact_type` and not `_type`. This way, the test passes. @raubitsj do you have any intel on this?

File: tests/test_redir.py (hunk 7 above)
Comment (dmitryduev) [reply tests/test_redir.py:234]: these were in the tests before...I assumed it's due to potentially slow i/o on the CI to make sure all those tens of thousands of lines of rubbish are picked up.

File: wandb/sdk/wandb_setup.py (hunk 8 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_setup.py:198]: ah, this came up in `https://wandb.atlassian.net/browse/WB-7410` and is actually what this whole refactoring thing originated from! the user's environment was messed up, which should not be an issue anymore with the new settings (things would either be ok or error out earlier + since now run.settings is exposed, it's much easier to debug). 
why needing two if clauses: if _server is None, try setting it up, and if that that still fails, don't do anything. the hope is that this will now never happen :)

File: wandb/sdk/wandb_init.py (hunk 10 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_init.py:229]: ah, this is one of outstanding questions I have and I am hoping to get input from @raubitsj on it. Why bother with freezing at all if everything is now supposed to be consistent, regardless of the order in which individual settings from different sources are loaded?

File: wandb/sdk/wandb_setup.py (hunk 8 above)
Comment (kptkin) [reply wandb/sdk/wandb_setup.py:198]: Ah I see now... makes sense... should we leave a comment here?

File: wandb/sdk/wandb_init.py (hunk 12 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_init.py:115]: Ah, I want my precious reviewers to double-check the logic re policy/non-policy settings and their source priorities pleeease :)

Comment (kptkin) [reply wandb/sdk/wandb_init.py:115]: Ah, haha... task for the reader ;)

File: requirements_dev.txt [hunk 15]
```diff
@@ -31,6 +31,7 @@ jax[cpu]; python_version < '3.10'
 fastcore; python_version < '3.10'
 pyarrow; python_version < '3.10'
 metaflow>=2.3.5; python_version < '3.10'
+transformers; python_version < '3.10'
```
Comment (raubitsj) [inline requirements_dev.txt:34]: Why is this needed?  i dont think we use transformers in pytest env.. (i think we were mocking the module only)

File: tests/integrations/test_keras.py (hunk 5 above)
Comment (raubitsj) [reply tests/integrations/test_keras.py:86]: well.. technically it is relying on `wandb_init_run` which is also doing the init for you, i think it makes sense that if you are using a fixture to do the init you can assume it will do the finish for you.

File: requirements_dev.txt (hunk 15 above)
Comment (dmitryduev) [reply requirements_dev.txt]: Yea, my bad - I screwed up indentation 🙈 
fixed here: https://github.com/wandb/client/pull/3083/commits/c939dcbb32391b58e710df6505f07631638eb61e

File: tests/test_data_types.py [hunk 16]
```diff
@@ -455,79 +457,82 @@ def test_video_path_invalid(runner):
...
         wandb.Molecule.from_rdkit(mol_file_name)
+    wandb.finish()
 
 
 def test_html_str(mocked_run):
     html = wandb.Html("<html><body><h1>Hello</h1></body></html>")
     html.bind_to_run(mocked_run, "rad", "summary")
     wandb.Html.seq_to_json([html], mocked_run, "rad", "summary")
     assert os.path.exists(html._path)
+    wandb.finish()
```
Comment (vanpelt) [inline tests/test_data_types.py:535]: I'm curious why we added all of these?  The `mocked_run` object should be calling finish after yielding it to the test.

File: tests/wandb_agent_test.py [hunk 17]
```diff
@@ -55,6 +58,7 @@ def train():
         run = wandb.init(entity="ign", project="ignored")
         sweep_projects.append(run.project)
         sweep_entities.append(run.entity)
+        run.finish()
```
Comment (vanpelt) [inline tests/wandb_agent_test.py:61]: Same here, I would think the sweep agents are automatically calling `finish` when the function returns...

File: tests/wandb_run_test.py [hunk 18]
```diff
@@ -104,6 +134,7 @@ def test_log_code(test_settings):
         f.write("Not that big")
     art = run.log_code()
     assert sorted(art.manifest.entries.keys()) == ["test.py"]
+    run.finish()
```
Comment (vanpelt) [inline tests/wandb_run_test.py:137]: Same here, the `test_settings` fixture has:

```
if wandb.run is not None:
        wandb.run.finish()
```

I'm sure there was a reason for this, but if we can make it "Just Work" in tests it's going to make future devs lives a lot less error prone.

File: wandb/sdk/wandb_settings.py [hunk 19]
```diff
@@ -165,56 +112,32 @@ def _get_program_relpath_from_gitrepo(
...
-    """Redact a dict of unsafe values specified by their key."""
-    if not d or unsafe_keys.isdisjoint(d):
-        return d
-    safe_dict = d.copy()
-    safe_dict.update({k: redact_str for k in unsafe_keys.intersection(d)})
-    return safe_dict
+@enum.unique
+class Source(enum.IntEnum):
+    OVERRIDE: int = 0
+    BASE: int = 1  # fixme: audit this
```
Comment (vanpelt) [inline wandb/sdk/wandb_settings.py:128]: Would love to know a little more about this comment ;)

File: wandb/sdk/wandb_settings.py [hunk 20]
```diff
@@ -223,925 +146,1015 @@ class SettingsConsole(enum.Enum):
...
+            "hook": lambda x: (
+                self._path_convert(
+                    self.wandb_dir, f"{self.run_mode}-{self.timespec}-{self.run_id}", x
+                )
+                or tempfile.gettempdir()
+            ),
+        }
+        self.username: Any = {
+            "validator": lambda x: isinstance(x, str),
+        }
```
Comment (vanpelt) [inline wandb/sdk/wandb_settings.py:766]: All of these `Any` annotations make me feel sad.  Any reason not to make these NamedTuples or something with strong type gurarentees?

File: wandb/sdk/wandb_settings.py [hunk 21]
```diff
@@ -223,925 +146,1015 @@ class SettingsConsole(enum.Enum):
...
-                    raise KeyError(k)
-                v = self._perform_preprocess(k, check[k])
-                self._check_invalid(k, v)
-                data[k] = v
-        for k, v in six.iteritems(data):
-            if v is None:
-                continue
-            if self._priority_failed(k, source=_source, override=_override):
+        env_prefix: str = "WANDB_"
+        special_env_var_names = {
```
Comment (vanpelt) [inline wandb/sdk/wandb_settings.py:989]: Would be good to add a note about what makes these specials and when a dev should add something to this list.

**Review by vanpelt — APPROVED** [review]
Overall looks great!  One addition I would suggest is a quick explainer on how to add new settings.  We could put this in a comment of wandb_settings or one of our doc files, but there's a lot to think about between env vars, validation, and serialization that would be good to clearly define.

File: wandb/sdk/wandb_settings.py (hunk 20 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_settings.py:766]: All these are converted to `Property` class instances immediately after definition, all of which come with validators that guarantee not only types, but also anything else, specific to each setting (see e.g. https://github.com/wandb/client/pull/3083/files/cf4a6864f7e306dd5d3fc1172fb26f29fc96265f#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R476). Here's my reasoning for doing it this way:
- Properties encapsulate all the logic we want from them: preprocessing, validation, runtime hooks, deps, source priorities (see an example in the pr description). Yes, static checkers aren't of much use in this case, action happens at runtime. But we should bring our test coverage to ~100% anyways :) 
  - Validation allows programming by contract of sorts (long live [PEP316](https://www.python.org/dev/peps/pep-0316/) ^_^), which static checkers won't help with.
- I could have gone with `object.__setattr__` to define things (like we're doing now) but that breaks code completion in pycharm (and i'd expect same in vscode), and it's kind of nice to have.
- We could define these as Properties straight away:
```
self.mofl: Property = Property(
     name="mofl",
     validator=[lambda x: isinstance(x, (int, float)), lambda x: x > 42],
 )
```
or something, it just takes up more space to spell it out since I decided to attach names to them. I'm afraid it might freak mypy out, although I haven't tried to be honest.

**Comment by dmitryduev** [comment]
Many thanks @vanpelt!

> Overall looks great! One addition I would suggest is a quick explainer on how to add new settings. We could put this in a comment of wandb_settings or one of our doc files, but there's a lot to think about between env vars, validation, and serialization that would be good to clearly define.
Good point, will add a section to `CONTRIBUTING.md`, there's quite a bit going on to have it only in the comments, me thinks.

File: wandb/sdk/wandb_settings.py (hunk 19 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_settings.py:128]: https://youtu.be/HQ47glxcxr0?t=13 ;)

It's a general comment that we should use these more or ditch - e.g. ENTITY or WORKSPACE never come up in the code, ORG only come up in tests, etc.

File: tests/wandb_run_test.py (hunk 18 above)
Comment (dmitryduev) [reply tests/wandb_run_test.py:137]: Yea, I was trying to find the reason for flakiness and the standing hypothesis was that maybe for some reason the singleton doesn't get recycled properly or something, which is of course not true since all fixtures are func-scoped.
I later discovered that it's (mainly) because of stuff leaking into `os.environ` and the filesystem through certain fixtures and tests, but decided to keep these calls in as it seemed more explicit and a settings fixture didn't seem like the first place I'd look for action like that. 


Now that you bring this up, how about ditching it from `test_settings` and instead doing something like
```python
@pytest.fixture(autouse=True)
def cleanup():
    yield
    if wandb.run is not None:
        wandb.run.finish()
```
What do you think? @raubitsj?

File: wandb/sdk/wandb_settings.py (hunk 20 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_settings.py:766]: @raubitsj has a great idea about how to get the best of both static and runtime worlds, implementing now.
[the idea is to use typed class attributes + dynamic re-definition as properties propagating type annotation as the first validator in the resulting property object; this does add a bit of pain to the dev experience, but I'm adding a note to the contrib guide + a unit test that ensures people don't screw up in the process].

**Review by raubitsj — APPROVED** [review]
Awesome. Lots of great stuff here!
